tfx_bsl/beam/run_inference.py

```diff
@@ -221,17 +232,207 @@ def update(self, elements: List[Union[tf.train.Example,
+    self.full_model_name = name_spec.format(project_id, model_name,
+                                            version_name)
+
+  def setup(self):
+    super(_RemotePredictDoFn, self).setup()
+    self._api_client = self._get_api_client()
+
+  @staticmethod
+  def _get_api_client() -> discovery.Resource:
+    return discovery.build('ml', 'v1')
+
+  @staticmethod
+  def _execute_request(request: http.HttpRequest) -> dict:
+    return request.execute()
+
+  def _make_request(self, model_name: str, body: dict) -> http.HttpRequest:
+    return self._api_client.projects().predict(name=model_name, body=body)
+
+  @classmethod
+  def _prepare_instances(cls,
+    elements: List[Union[tf.train.Example, tf.train.SequenceExample]]
+  ) -> Generator[Mapping[Text, Any], None, None]:
+    for example in elements:
+      # TODO: support tf.train.SequenceExample
+      if not isinstance(example, tf.train.Example):
+        raise ValueError('Remote prediction only supports tf.train.Example')
+
+      instance = {}
+      for input_name, feature in example.features.feature.items():
+        attr_name = feature.WhichOneof('kind')
+        attr = getattr(feature, attr_name)
+        values = cls._parse_feature_content(attr.value, attr_name,
+                                            cls._sending_as_binary(input_name))
+        values = cls._maybe_flatten(values)
+        instance[input_name] = values
+      yield instance
+
+  @staticmethod
+  def _sending_as_binary(input_name: Text) -> bool:
+    """Whether data should be sent as binary."""
+    return input_name.endswith('_bytes')
+
+  @staticmethod
+  def _parse_feature_content(
+      values: Iterable[Any], attr_name: Text, as_binary: bool
+  ) -> Sequence[Any]:
+    """Parse the content of tf.train.Feature object.
+
+    If bytes_list, parse a list of bytes-like objects to a list of strings so
+    that it would be JSON serializable.
+
+    If float_list or int64_list, do nothing.
+
+    If data should be sent as binary, mark it as binary by replacing it with
+    a single attribute named 'b64'.
+    """
+    if as_binary:
+      values = [{'b64': base64.b64encode(x).decode()} for x in values]
+    elif attr_name == 'bytes_list':
+      values = [x.decode() for x in values]
+    else:
+      values = [x for x in values]
+    return values
+
+  @staticmethod
+  def _maybe_flatten(values: Sequence[Any]) -> Union[Sequence[Any], Any]:
+    """Flatten a sequence if its length is 1."""
+    if len(values) == 1:
+      result = values[0]
+    else:
+      result = values
+    return result
+
+  def run_inference(
+      self, elements: List[Union[tf.train.Example, tf.train.SequenceExample]]
+  ) -> Sequence[Mapping[Text, Any]]:
+    body = {
+      'instances': list(self._prepare_instances(elements))
+    }
+    response = self._execute_request(self._make_request(self.full_model_name,
+                                                        body))
+    try:
+      error_msg = response['error']
```

Comment:
Is response a map? If so, you can use:
if 'error' in response

Reply:
Done. It looks a bit better after that the change.